server/src/main/java/io/deephaven/server/arrow/ExchangeMarshallerModule.java

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright (c) 2016-2025 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.server.arrow;
+
+import dagger.Module;
+import dagger.Provides;
+import dagger.multibindings.ElementsIntoSet;
+import io.deephaven.extensions.barrage.BarrageMessageWriter;
+import io.deephaven.server.session.SessionService;
+import io.deephaven.server.util.Scheduler;
+
+import java.util.*;
+
+/**
+ * A dagger module that provides {@link ExchangeMarshaller} s for use by the {@link ArrowFlightUtil}
+ * {@code DoExchangeMarshaller}, loaded using a {@link ServiceLoader} constructed with the injected {@link Scheduler},
+ * {@link io.deephaven.server.session.SessionService.ErrorTransformer} and {@link BarrageMessageWriter.Factory}
+ * parameters.
+ */
+@Module
+public class ExchangeMarshallerModule {
+    @Provides
+    @ElementsIntoSet
+    public static Set<ExchangeMarshaller> provideExchangeMarshallers(final Scheduler scheduler,
+            final SessionService.ErrorTransformer errorTransformer,
+            final BarrageMessageWriter.Factory streamGeneratorFactory) {
+        final Iterator<Factory> it = ServiceLoader.load(ExchangeMarshallerModule.Factory.class).iterator();
+        if (!it.hasNext()) {
+            return Collections.emptySet();
+        }
+
+        final List<ExchangeMarshaller> list = new ArrayList<>();
+
+        while (it.hasNext()) {
+            list.add(it.next().create(scheduler, errorTransformer, streamGeneratorFactory));
+        }
+
+        // Note that although we sort the marshallers by priority, the eventual user of the marshaller must sort the
+        // complete set by priority.
+        list.sort(Comparator.comparingInt(ExchangeMarshaller::priority));
+
+        return Collections.unmodifiableSet(new LinkedHashSet<>(list));
+    }
+
+    /**
+     * To add an additional {@link ExchangeMarshaller}, implement this Factory and add it as a service.
+     */
+    public interface Factory {
+        ExchangeMarshaller create(final Scheduler scheduler,
+                final SessionService.ErrorTransformer errorTransformer,
+                final BarrageMessageWriter.Factory streamGeneratorFactory);
+    }
+
+    @Provides
+    @ElementsIntoSet
+    public static Set<ExchangeRequestHandlerFactory> provideRequestHandlers() {
+        final Iterator<ExchangeRequestHandlerFactory> it =
+                ServiceLoader.load(ExchangeRequestHandlerFactory.class).iterator();
+        if (!it.hasNext()) {
+            return Collections.emptySet();
+        }
```

Comment:
nitpick: this seems silly. dh as a server would be useless if there was no registered factory (here and above), and it costs us nothing except lines of code to make the extra arraylist/set below in the cases where a dh server has no data to serve...

Reply:
Did you mean to resolve this?

I'd rather the streaming above go back to iteration or this iteration change to streaming so they are consistent.

Reply:
No, tunnel vision thinking it was about the first one.